store/src/errors.rs

```diff
@@ -1,64 +1,198 @@
+use rusqlite::types::FromSqlError;
+use thiserror::Error;
+use tokio::sync::oneshot::error::RecvError;
+
+// Conversion errors
+// =================================================================================================
+
+#[derive(Error, Debug)]
+pub enum ConversionError {
+    #[error("Parse error: {0}")]
+    ParseError(#[from] ParseError),
+    #[error("Field `{field_name}` required to be filled in protobuf representation of {entity}")]
+    MissingFieldInProtobufRepresentation {
+        entity: &'static str,
+        field_name: &'static str,
+    },
+}
+
+// Error type from Deadpool's `interaction` task errors
+// =================================================================================================
+
+#[derive(Debug, Error)]
+pub enum InteractionTaskError {
+    #[error("Migration task failed: {0}")]
+    MigrationTaskFailed(String),
+    #[error("Select nullifiers task failed: {0}")]
+    SelectNullifiersTaskFailed(String),
+    #[error("Select notes task failed: {0}")]
+    SelectNotesTaskFailed(String),
+    #[error("Select accounts task failed: {0}")]
+    SelectAccountsTaskFailed(String),
+    #[error("Select block header task failed: {0}")]
+    SelectBlockHeaderTaskFailed(String),
+    #[error("Select block headers task failed: {0}")]
+    SelectBlockHeadersTaskFailed(String),
+    #[error("Select account hashes task failed: {0}")]
+    SelectAccountHashesTaskFailed(String),
+    #[error("Get state sync task failed: {0}")]
+    GetStateSyncTaskFailed(String),
+    #[error("Apply block task failed: {0}")]
+    ApplyBlockTaskFailed(String),
+}
+
+// Database errors
+// =================================================================================================
+
+#[derive(Debug, Error)]
+pub enum DbError {
+    #[error("Missing database connection: {0}")]
+    MissingDbConnection(#[from] PoolError),
+    #[error("SQLite error: {0}")]
+    SqliteError(#[from] rusqlite::Error),
+    #[error("SQLite error: {0}")]
+    FromSqlError(#[from] FromSqlError),
+    #[error("SQLite migration error: {0}")]
+    SqliteMigrationError(#[from] rusqlite_migration::Error),
+    #[error("Pool build error: {0}")]
+    PoolBuildError(#[from] deadpool_sqlite::BuildError),
+    #[error("I/O error: {0}")]
+    IoError(#[from] io::Error),
+    #[error("Prost decode error: {0}")]
+    DecodeError(#[from] DecodeError),
+    #[error("SQLite pool interaction task failed: {0}")]
+    InteractionTaskError(#[from] InteractionTaskError),
+    #[error("Genesis block error: {0}")]
+    GenesisBlockError(#[from] GenesisError),
+    #[error("State error: {0}")]
+    StateError(Box<StateError>),
+    #[error("Conversion error: {0}")]
+    ConversionError(#[from] ConversionError),
+    #[error("Decoding nullifier from database failed: {0}")]
+    NullifierDecodingError(DeserializationError),
+    #[error("Block applying was broken because of closed channel on state side: {0}")]
+    ApplyBlockFailedClosedChannel(RecvError),
+}
```

Comment:
Let's rename this error type to `DatabaseError`. Also, lets remove unused variants (e.g., I think `StateError` variant is not used).

In generally, would be great to make this error specifically about interacting with the database and reduce the number of variants (e.g., I don't think `GenesisBlockError` should be here, but that'll require a bigger refactoring). Let's create an issue for this.

Reply:
@bobbinth I've extracted several variants into a new `DatabaseSetupError` class and slightly refactored `ensure_genesis_block` function.